Add Control Flow Guard support

alvinhochun opened the pull request:
Adds the option to build with Control Flow Guard support. This is disabled by default and currently only enabled for the scheduled GitHub Actions build because it requires latest LLVM 16 and mingw-w64. We should change the default to enabled for the next release.

Closes https://github.com/mstorsjo/llvm-mingw/issues/301

mstorsjo commented on run-tests.sh:
```diff
@@ -42,6 +42,16 @@ if ! $ANY_ARCH-w64-mingw32-gcc -E is-ucrt.c > /dev/null 2>&1; then
 rm -f is-ucrt.c
 
+echo "int main(){}" | $ANY_ARCH-w64-mingw32-clang -x c++ - -o has-cfguard-test.exe -mguard=cf
+if llvm-readobj --coff-load-config has-cfguard-test.exe | grep -q 'CF_INSTRUMENTED (0x100)'; then
```
If I understand correctly - this check is only capable of discerning whether the compiler actually did handle `-mguard=cf` by setting the instrumentation flags on the main object file, while the rest of the toolchain might have been built without `--enable-cfguard`?

What happens with older versions of clang that don't recognize `-mguard=cf`, won't they error out when executing that clang command with an unrecognized option?

alvinhochun replied:
The check should fail when mingw-w64-crt isn't built with `--enable-cfguard` because `_load_config_used` will be missing and then the CF flags won't exist.

You're right, older clang would error out on `-mguard=cf` so I will need to wrap this in another `if` block too.

mstorsjo commented on run-tests.sh:
```diff
@@ -217,6 +233,14 @@ for arch in $ARCHS; do
+            $arch-w64-mingw32-clang $test.c -o $TEST_DIR/$test-asan-cfguard.exe -fsanitize=address -g -gcodeview -Wl,-pdb,$TEST_DIR/$test-asan.pdb -mguard=cf
+            if [ -n "$NATIVE" ]; then
+                TESTS_EXTRA="$TESTS_EXTRA $test-asan-cfguard"
+            FAILURE_TESTS="$FAILURE_TESTS $test-asan-cfguard"
```
This line seems incorrectly indented

mstorsjo commented on run-tests.sh:
```diff
@@ -331,6 +355,22 @@ for arch in $ARCHS; do
+                    # (STATUS_STACK_BUFFER_OVERRUN aka fail fast exception).
+                    # MSYS2 bash does not give us the full 32-bit exit code, so
+                    # we have to rely on cmd.exe to perform the check.
+                    # (This probably doesn't work on WINE, but WINE doesn't
```
Nit; I'd prefer to spell Wine with regular upper/lower casing instead of all-uppercase.

mstorsjo commented on run-tests.sh:
```diff
@@ -331,6 +355,22 @@ for arch in $ARCHS; do
+                    # we have to rely on cmd.exe to perform the check.
+                    # (This probably doesn't work on WINE, but WINE doesn't
+                    # support CFG anyway, at least not for now...)
+                    $RUN cmd //c "$test.exe invalid_icall & if errorlevel -1073740791 (if not errorlevel -1073740790 (exit 0)) & exit 1"
```
What's the deal with the second check `if not errorlevel -1073740790` here? The former check, `-1073740791`, corresponds to the expected `0xc0000409`, but what's the second one?

alvinhochun replied:
`if errorlevel N` is effectively `if errorlevel >= N`, so I need to do 'errorlevel >= N and not errorlevel >= N+1' to check for equality. (I can't check equality of `%errorlevel%` directly because of how cmd's variable substitution works.)

mstorsjo replied:
Ah, I see. That's rather nonobvious and clearly worthy of a comment!

alvinhochun replied:
Oh, maybe I actually can use a more direct equality comparison if I enable delayed expansion, let me try that.

mstorsjo commented on test/cfguard-test.c:
```diff
@@ -0,0 +1,102 @@
+#define WINVER 0x0603
```
Please add a license header to this file

mstorsjo commented on test/cfguard-test.c:
```diff
@@ -0,0 +1,102 @@
+        return 0;
+    if (policy.EnableControlFlowGuard)
+        return 1;
+    return 0;
```
Nitpick; what about just `return policy.EnableControlFlowGuard;` here?

mstorsjo reviewed (commented):
Looking extremely good overall, thanks! I've only got a couple minor details to discuss.

alvinhochun commented:
Thanks, I have addressed the comments. Let's wait and see how the actions https://github.com/alvinhochun/llvm-mingw/actions/runs/3158626818 and https://github.com/alvinhochun/llvm-mingw/actions/runs/3158634041 go.

mstorsjo commented:
> Thanks, I have addressed the comments. Let's wait and see how the actions https://github.com/alvinhochun/llvm-mingw/actions/runs/3158626818 and https://github.com/alvinhochun/llvm-mingw/actions/runs/3158634041 go.

Thanks! Can you kick off another actions run, with a branch on top of this which bumps llvm/mingw to versions that support it, and which enables it by default even for non-scheduled runs, to see it in action?

alvinhochun commented:
> > Thanks, I have addressed the comments. Let's wait and see how the actions https://github.com/alvinhochun/llvm-mingw/actions/runs/3158626818 and https://github.com/alvinhochun/llvm-mingw/actions/runs/3158634041 go.
> 
> Thanks! Can you kick off another actions run, with a branch on top of this which bumps llvm/mingw to versions that support it, and which enables it by default even for non-scheduled runs, to see it in action?

https://github.com/alvinhochun/llvm-mingw/actions/runs/3158634041 does use the latest versions of llvm and mingw-w64 (the get-versions version check is enabled) and has cfguard enabled.

mstorsjo commented:
> > > Thanks, I have addressed the comments. Let's wait and see how the actions https://github.com/alvinhochun/llvm-mingw/actions/runs/3158626818 and https://github.com/alvinhochun/llvm-mingw/actions/runs/3158634041 go.
> > 
> > 
> > Thanks! Can you kick off another actions run, with a branch on top of this which bumps llvm/mingw to versions that support it, and which enables it by default even for non-scheduled runs, to see it in action?
> 
> https://github.com/alvinhochun/llvm-mingw/actions/runs/3158634041 does use the latest versions of llvm and mingw-w64 (the get-versions version check is enabled) and has cfguard enabled.

These test runs seem to have run successfully - so, I guess there's nothing left stopping this from getting merged then?

alvinhochun commented:
> These test runs seem to have run successfully - so, I guess there's nothing left stopping this from getting merged then?

Not from me!

mstorsjo commented:
> > These test runs seem to have run successfully - so, I guess there's nothing left stopping this from getting merged then?
> 
> Not from me!

Great! I did some extra test runs with Wine (with run-tests.sh tweaked to actually run the failure tests there too), and it worked as expected (`check_enabled` indicated that it wasn't enabled and skipped the rest), so this looks good to go to me. Thanks!
